File: app/src/net/reichholf/dreamdroid/helpers/enigma2/Timer.java
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 * 
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.helpers.enigma2;

import android.app.Activity;
import android.content.Intent;
import android.support.v4.app.Fragment;

import net.reichholf.dreamdroid.DreamDroid;
import net.reichholf.dreamdroid.R;
import net.reichholf.dreamdroid.activities.SimpleToolbarFragmentActivity;
import net.reichholf.dreamdroid.activities.abs.MultiPaneHandler;
import net.reichholf.dreamdroid.fragment.TimerEditFragment;
import net.reichholf.dreamdroid.helpers.DateTime;
import net.reichholf.dreamdroid.helpers.ExtendedHashMap;
import net.reichholf.dreamdroid.helpers.NameValuePair;
import net.reichholf.dreamdroid.helpers.Statics;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

/**
 * @author sreichholf
 */
public class Timer {
	public static final String DATA = "data";

	public static final String KEY_REFERENCE = "reference";
	public static final String KEY_SERVICE_NAME = "servicename";
	public static final String KEY_EIT = "eit";
	public static final String KEY_NAME = "name";
	public static final String KEY_DESCRIPTION = "description";
	public static final String KEY_DESCRIPTION_EXTENDED = "descriptionex";
	public static final String KEY_DISABLED = "disabled";
	public static final String KEY_BEGIN = "begin";
	public static final String KEY_BEGIN_READEABLE = "begin_readable";
	public static final String KEY_END = "end";
	public static final String KEY_END_READABLE = "end_readable";
	public static final String KEY_DURATION = "duration";
	public static final String KEY_DURATION_READABLE = "duration_readable";
	public static final String KEY_START_PREPARE = "startprepare";
	public static final String KEY_JUST_PLAY = "justplay";
	public static final String KEY_AFTER_EVENT = "afterevent";
	public static final String KEY_LOCATION = "location";
	public static final String KEY_TAGS = "tags";
	public static final String KEY_LOG_ENTRIES = "logentries";
	public static final String KEY_FILE_NAME = "filename";
	public static final String KEY_BACK_OFF = "backoff";
	public static final String KEY_NEXT_ACTIVATION = "nextactivation";
	public static final String KEY_FIRST_TRY_PREPARE = "firsttryprepare";
	public static final String KEY_STATE = "state";
	public static final String KEY_REPEATED = "repeated";
	public static final String KEY_DONT_SAVE = "dontsave";
	public static final String KEY_CANCELED = "canceled";
	public static final String KEY_TOGGLE_DISABLED = "toggledisabled";

	public enum TimerStates {
		WAITING(0),
		PREPARED(1),
		RUNNING(2),
		ENDED(3);

		private int value;

		TimerStates(int val) {
			value = val;
		}

		@Override
		public String toString() {
			return String.valueOf(value);
		}

		public int intValue() {
			return value;
		}

		public String getText(Activity ac) {
			return (String) ac.getResources().getTextArray(R.array.afterevents)[value];
		}
	}

	public enum Afterevents {
		NOTHING(0),
		STANDBY(1),
		DEEP_STANDBY(2),
		AUTO(3);

		private int value;

		Afterevents(int val) {
			value = val;
		}

		@Override
		public String toString() {
			return String.valueOf(value);
		}

		public int intValue() {
			return value;
		}

		public String getText(Activity ac) {
			return (String) ac.getResources().getTextArray(R.array.afterevents)[value];
		}
	}

	/**
	 * @return
	 */
	public static ExtendedHashMap getInitialTimer() {
		ExtendedHashMap timer = new ExtendedHashMap();
		timer.put(KEY_DESCRIPTION, "");
		timer.put(KEY_LOCATION, "/hdd/movie/");
		timer.put(KEY_DISABLED, "0"); // enabled
		timer.put(KEY_JUST_PLAY, "0"); // record
		timer.put(KEY_AFTER_EVENT, Afterevents.AUTO.toString()); // auto
		timer.put(KEY_REPEATED, "0"); // One-Time-Event

		// Calculate values for begin and end
		Calendar cal = GregorianCalendar.getInstance();
		cal.setTime(new Date());

		long s = (cal.getTimeInMillis() / 1000);
		long e = s + 3600;

		String begin = Long.toString(s);
		String end = Long.toString(e);

		timer.put(KEY_BEGIN, begin);
		timer.put(KEY_END, end);

		return timer;
	}

	/**
	 * @param event
	 * @return
	 */
	public static ExtendedHashMap createByEvent(ExtendedHashMap event) {
		ExtendedHashMap timer = getInitialTimer();

		String start = event.getString(Event.KEY_EVENT_START);
		int duration = DateTime.parseTimestamp(event.getString(Event.KEY_EVENT_DURATION));
		int end = duration + DateTime.parseTimestamp(start);

		timer.put(Timer.KEY_BEGIN, start);
		timer.put(Timer.KEY_END, String.valueOf(end));
		timer.put(Timer.KEY_NAME, event.getString(Event.KEY_EVENT_TITLE));
		timer.put(Timer.KEY_DESCRIPTION, event.getString(Event.KEY_EVENT_DESCRIPTION));
		timer.put(Timer.KEY_DESCRIPTION_EXTENDED, event.getString(Event.KEY_EVENT_DESCRIPTION_EXTENDED));
		timer.put(Timer.KEY_SERVICE_NAME, event.getString(Event.KEY_SERVICE_NAME));
		timer.put(Timer.KEY_REFERENCE, event.getString(Event.KEY_SERVICE_REFERENCE));

		return timer;
	}

	public static ArrayList<NameValuePair> getSaveParams(ExtendedHashMap timer, ExtendedHashMap timerOld) {
		/*
		 * URL to create /web/timerchange? sRef= &begin= &end= &name=
		 * &description= &dirname= &tags= &eit= &disabled= &justplay=
		 * &afterevent= &repeated= &channelOld= &beginOld= &endOld=
		 * &deleteOldOnSave=
		 */
		// Build Parameters using timer HashMap
		ArrayList<NameValuePair> params = new ArrayList<>();

		params.add(new NameValuePair("sRef", timer.getString(KEY_REFERENCE)));
		params.add(new NameValuePair("begin", timer.getString(KEY_BEGIN)));
		params.add(new NameValuePair("end", timer.getString(KEY_END)));
		params.add(new NameValuePair("name", timer.getString(KEY_NAME)));
		params.add(new NameValuePair("description", timer.getString(KEY_DESCRIPTION)));
		params.add(new NameValuePair("dirname", timer.getString(KEY_LOCATION)));
		params.add(new NameValuePair("tags", timer.getString(KEY_TAGS)));
		params.add(new NameValuePair("eit", timer.getString(KEY_EIT)));
		params.add(new NameValuePair("disabled", timer.getString(KEY_DISABLED)));
		params.add(new NameValuePair("justplay", timer.getString(KEY_JUST_PLAY)));
		params.add(new NameValuePair("afterevent", timer.getString(KEY_AFTER_EVENT)));
		params.add(new NameValuePair("repeated", timer.getString(KEY_REPEATED)));

		if (timerOld != null) {
			params.add(new NameValuePair("channelOld", timerOld.getString(KEY_REFERENCE)));
			params.add(new NameValuePair("beginOld", timerOld.getString(KEY_BEGIN)));
			params.add(new NameValuePair("endOld", timerOld.getString(KEY_END)));
			params.add(new NameValuePair("deleteOldOnSave", "1"));
		} else {
			params.add(new NameValuePair("deleteOldOnSave", "0"));
		}

		return params;
	}

	public static ArrayList<NameValuePair> getEventIdParams(ExtendedHashMap event) {
		ArrayList<NameValuePair> params = new ArrayList<>();

		params.add(new NameValuePair("sRef", event.getString(Event.KEY_SERVICE_REFERENCE)));
		params.add(new NameValuePair("eventid", event.getString(Event.KEY_EVENT_ID)));

		return params;
	}

	public static ArrayList<NameValuePair> getDeleteParams(ExtendedHashMap timer) {
		// URL - web/timerdelete?sRef=&begin=&end=
		ArrayList<NameValuePair> params = new ArrayList<>();

		params.add(new NameValuePair("sRef", timer.getString(KEY_REFERENCE)));
		params.add(new NameValuePair("begin", timer.getString(KEY_BEGIN)));
		params.add(new NameValuePair("end", timer.getString(KEY_END)));

		return params;
	}


	/**
	 * @param mph    - A MultiPaneHandler instance
	 * @param event  - A timer (ExtendedHashMap)
	 * @param target - The target fragment (after saving/cancellation)
	 */
	public static void editUsingEvent(MultiPaneHandler mph, ExtendedHashMap event, Fragment target) {
		Timer.edit(mph, Timer.createByEvent(event), target, true);
	}

	/**
	 * @param mph    - A MultiPaneHandler instance
	 * @param timer  - A timer (ExtendedHashMap)
	 * @param target - The target fragment (after saving/cancellation)
	 * @param create - set to true if a new timer should be created instead of editing an existing one
	 */
	public static void edit(MultiPaneHandler mph, ExtendedHashMap timer, Fragment target, boolean create) {
		ExtendedHashMap data = new ExtendedHashMap();
		data.put("timer", timer);
		data.put("action", create ? DreamDroid.ACTION_CREATE : Intent.ACTION_EDIT);

		Intent intent = new Intent(target.getContext(), SimpleToolbarFragmentActivity.class);
		intent.putExtra("fragmentClass", TimerEditFragment.class);
		intent.putExtra("titleResource", create ? R.string.new_timer : R.string.edit_timer);
		//intent.putExtra("menuResource", R.menu.save);
		intent.putExtra("serializableData", (Serializable) data);
		target.getActivity().startActivityForResult(intent, Statics.REQUEST_EDIT_TIMER);
	}
}

File: app/src/net/reichholf/dreamdroid/helpers/enigma2/Timer.kt
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 *
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.helpers.enigma2

import android.app.Activity
import java.util.Date
import java.util.GregorianCalendar
import net.reichholf.dreamdroid.R
import net.reichholf.dreamdroid.activities.MainActivity
import net.reichholf.dreamdroid.activities.abs.MultiPaneHandler
import net.reichholf.dreamdroid.enigma.Event
import net.reichholf.dreamdroid.enigma.Timer as TypedTimer
import net.reichholf.dreamdroid.helpers.DateTime
import net.reichholf.dreamdroid.helpers.NameValuePair

/**
 * Timer XML field names, after-event enums, and request helpers. UI uses [TypedTimer].
 */
class Timer {
    enum class TimerStates(private val value: Int) {
        WAITING(0),
        PREPARED(1),
        RUNNING(2),
        ENDED(3);

        override fun toString(): String = value.toString()

        fun intValue(): Int = value

        fun getText(ac: Activity): String =
            ac.resources.getTextArray(R.array.afterevents)[value] as String
    }

    enum class Afterevents(private val value: Int) {
        NOTHING(0),
        STANDBY(1),
        DEEP_STANDBY(2),
        AUTO(3);

        override fun toString(): String = value.toString()

        fun intValue(): Int = value

        fun getText(ac: Activity): String =
            ac.resources.getTextArray(R.array.afterevents)[value] as String
    }

    companion object {
        const val DATA: String = "data"

        const val KEY_REFERENCE: String = "reference"
        const val KEY_SERVICE_NAME: String = "servicename"
        const val KEY_EIT: String = "eit"
        const val KEY_NAME: String = "name"
        const val KEY_DESCRIPTION: String = "description"
        const val KEY_DESCRIPTION_EXTENDED: String = "descriptionex"
        const val KEY_DISABLED: String = "disabled"
        const val KEY_BEGIN: String = "begin"
        const val KEY_BEGIN_READEABLE: String = "begin_readable"
        const val KEY_END: String = "end"
        const val KEY_END_READABLE: String = "end_readable"
        const val KEY_DURATION: String = "duration"
        const val KEY_DURATION_READABLE: String = "duration_readable"
        const val KEY_START_PREPARE: String = "startprepare"
        const val KEY_JUST_PLAY: String = "justplay"
        const val KEY_AFTER_EVENT: String = "afterevent"
        const val KEY_LOCATION: String = "location"
        const val KEY_TAGS: String = "tags"
        const val KEY_LOG_ENTRIES: String = "logentries"
        const val KEY_FILE_NAME: String = "filename"
        const val KEY_BACK_OFF: String = "backoff"
        const val KEY_NEXT_ACTIVATION: String = "nextactivation"
        const val KEY_FIRST_TRY_PREPARE: String = "firsttryprepare"
        const val KEY_STATE: String = "state"
        const val KEY_REPEATED: String = "repeated"
        const val KEY_DONT_SAVE: String = "dontsave"
        const val KEY_CANCELED: String = "canceled"
        const val KEY_TOGGLE_DISABLED: String = "toggledisabled"

        fun getInitialTimer(): TypedTimer {
            val cal = GregorianCalendar.getInstance()
            cal.time = Date()
            val s = cal.timeInMillis / 1000
            val e = s + 3600
            return TypedTimer(
                description = "",
                location = "/hdd/movie/",
                disabled = "0",
                justPlay = "0",
                afterEvent = Afterevents.AUTO.toString(),
                repeated = "0",
                begin = s.toString(),
                end = e.toString()
            )
        }

        fun createByEvent(event: Event): TypedTimer {
            val duration = DateTime.parseTimestamp(event.duration)
            val end = duration + DateTime.parseTimestamp(event.start)
            return getInitialTimer().copy(
                begin = event.start,
                end = end.toString(),
                name = event.title,
                description = event.description,
                descriptionExtended = event.descriptionExtended,
                serviceName = event.serviceName,
                reference = event.serviceReference
            )
        }

        fun getSaveParams(timer: TypedTimer, timerOld: TypedTimer?): ArrayList<NameValuePair> {
            val params = ArrayList<NameValuePair>()
            params.add(NameValuePair("sRef", timer.reference))
            params.add(NameValuePair("begin", timer.begin))
            params.add(NameValuePair("end", timer.end))
            params.add(NameValuePair("name", timer.name))
            params.add(NameValuePair("description", timer.description))
            params.add(NameValuePair("dirname", timer.location))
            params.add(NameValuePair("tags", timer.tags))
            params.add(NameValuePair("eit", timer.eit))
            params.add(NameValuePair("disabled", timer.disabled))
            params.add(NameValuePair("justplay", timer.justPlay))
            params.add(NameValuePair("afterevent", timer.afterEvent))
            params.add(NameValuePair("repeated", timer.repeated))
            if (timerOld != null) {
                params.add(NameValuePair("channelOld", timerOld.reference))
                params.add(NameValuePair("beginOld", timerOld.begin))
                params.add(NameValuePair("endOld", timerOld.end))
                params.add(NameValuePair("deleteOldOnSave", "1"))
            } else {
                params.add(NameValuePair("deleteOldOnSave", "0"))
            }
            return params
        }

        fun getEventIdParams(event: Event): ArrayList<NameValuePair> {
            val params = ArrayList<NameValuePair>()
            params.add(NameValuePair("sRef", event.serviceReference))
            params.add(NameValuePair("eventid", event.eventId))
            return params
        }

        fun getDeleteParams(timer: TypedTimer): ArrayList<NameValuePair> {
            val params = ArrayList<NameValuePair>()
            params.add(NameValuePair("sRef", timer.reference))
            params.add(NameValuePair("begin", timer.begin))
            params.add(NameValuePair("end", timer.end))
            return params
        }

        fun editUsingEvent(mph: MultiPaneHandler?, event: Event) {
            edit(mph, createByEvent(event), true)
        }

        fun edit(mph: MultiPaneHandler?, timer: TypedTimer, create: Boolean) {
            (mph as? MainActivity)?.phoneNav?.navigateToTimerEdit(timer, create)
        }
    }
}
